Thecarisma-syntax.md

```diff
@@ -0,0 +1,620 @@
+
+- Calling a type method.
+
+    ```rust
+    const john = Person::new("John", 40)
+    ```
+
+#### Instance methods
+- Instances of a type can have associated functions. We call them `instance methods`.
+
+    ```swift
+    impl Person {
+        func get_name(self) -> String {
+            self.name
+        }
+    }
+    ```
+
+    NB: `self` here refers to the type instance.
+
+- Calling an instance method.
+
+    ```rust
+    john.get_name()
+
+    get_name(john)
+    ```
+
+
+#### Instantiation
+- You can instantiate types.
+    ```rust
+    type Person {
+        name: String,
+        age: Int
+    }
+    ```
+
+    ```rust
+    const person = Person::new("John", 45)
+    ```
+
+    ```
+    window = Window {
+        title: "Hello World",
+        extended: true
+    }
+    ```
+
+## UNIFORM FUNCTION CALL SYNTAX (UFCS)
+- UFCS is a feature that blurs the line between instance methods and regular functions.
+
+- A regular function can be called using both the dot notation call syntax or the regular function call syntax.
+
+    ```swift
+    func get_name(p: Person) -> String {
+        p.name
+    }
+
+    const person = Person::new("John", 52)
+
+    person.get_name() // dot notation syntax
+    get_name(person) // regular call syntax
+    ```
+
+- Likewise instance methods can be called using both the dot notation call syntax or the regular function call syntax.
+
+    ```swift
+    impl Person {
+        func get_name(self) -> String {
+            self.name
+        }
+    }
+
+    const person = Person::new("John", 52)
+
+    person.get_name() // dot notation syntax
+    get_name(person) //  regular call syntax
+    ```
+
+#### Why UFCS is useful
+
+Sometimes it makes sense to have certain functions be instance methods because they associate closely to a particular type. Other times it makes sense for such functions to be regular top-level functions because they don't associate with any single type. With UFCS, you don't have to memorize which one is a function or which one is an instance method, because you can call them using the both dot notation call and regular call syntaxes.
+
+
+- Take `map` function for example. It is a generic function and does not associate with any single type.
+
+    ```swift
+    func map[T, U](iter: T, fn: (U) -> U) -> T {
+        var new = T::new()
+        for (i, x) in enumerate(iter) {
+            new[i] = fn(x)
+        }
+        new
+    }
+    ```
+
+    Because we have UFCS support, we can chain up calls to regular function using the dot notation syntax, which can be easier to read sometimes.
+
+    ```swift
+    map(filter(map(arr, (x) => x * 2), (x) => x > 2), (x) => x * 2)
+    ```
+
+    ```swift
+    arr
+        .map((x) => x * 2)
+        .filter((x) => x > 2)
+        .map((x) => x * 2)
+    ```
+
+
+
+## TRAITS
+- Ratio does not support inheritance. Instead it choses composition-based type extension.
+
+    ```rust
+    trait Callable {
+        call: (...Int) -> Int
+    }
```

Comment:
I so much agree with the idea of adopting traits. 

Although, we'll have to take into consideration certain scenarios, and establish a means for resolution. E.g.
- Multiple traits with at least 1 similar function definition,
- Can traits inherit other traits?
- etc.